Page/fees explainer for users

platocrat opened the pull request:
<!--
Need help?
Refer to our contributing guidelines for additional information about making a good pull request:
https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**
First section for the User page!
Adding detailed "How Fees Work in Optimistic Ethereum" section.

**Metadata**
- Fixes #39

snario commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,236 @@
+We'll be primarily covering gas costs and savings experienced in _and_ around OE when you:
+
+1. Initiate an L1 Deposit
+2. Complete a withdrawal (i.e. L2 → L1)
```
```suggestion
2. Complete a withdrawal (i.e., L2 → L1)
```
https://english.stackexchange.com/questions/6894/should-there-be-a-comma-after-i-e

maurelian commented on src/.vuepress/config.js:
```diff
@@ -65,6 +65,17 @@ module.exports = {
+          ],
+          collapsable: false,
+          sidebarDepth: 1,
+        },
```
This comment pertains to line 83 below, and is unrelated to this PR. 

I had the impression that the `TLDR` page was removed. It's not on the current site, but it does appear in the [netlify preview link](https://deploy-preview-63--thirsty-swanson-df3e28.netlify.app/docs/) generated by the CI checks.

platocrat replied:
I think that's because this branch is forked from the branch that re-adds the TL;DR page.

See this snapshot below of the commits shown above, in this PR.

<img width="717" alt="Screen Shot 2021-04-08 at 3 38 34 PM" src="https://user-images.githubusercontent.com/37757724/114093251-7de00000-9880-11eb-9b98-388e295c156e.png">

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,236 @@
+Due to this slight difference, we recommend using `eth_estimateGas` to calculate your `gasLimit`.
+This is because `gasPrice` is a constant value of `1 Gwei`, so your focus should be on optimizing the `gasLimit`.
+
+Two critical points to remember for this section here are: 1) you should never drop your `eth_gasPrice` or your transaction will simply get rejected on L2, and 2) never drop your `eth_estimateGas` `gasLimit` or your transaction might revert.
```
> never drop your `eth_estimateGas` `gasLimit` or your transaction might revert.

Do you mean, 'never drop `gasLimit` below the value returned by `eth_estimateGas`?

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,236 @@
+#### Final note on `gasUsed`
+
+Some final notes on the amount of `gasUsed` per transaction on L2.
+* **The `gasUsed` must always be under 9 million**.
```
```suggestion
* **`gasUsed` must always be under 9 million per transaction**.
```

maurelian replied:
Similarly remove "The" from the following lines

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,236 @@
+
+Some final notes on the amount of `gasUsed` per transaction on L2.
+* **The `gasUsed` must always be under 9 million**.
+* **The `gasPrice` is fixed as a constant (it is initially set to `1 Gwei`)**
+* **The `gasUsed` is always equal to the `gasLimit` (since 100% of gas is always used)**
+* Lastly, **we add `executionGasUsed` as a new field in the transaction receipt.
```
```suggestion
* `gasPrice` is fixed as a constant (it is initially set to `1 Gwei`).
* `gasUsed` is always equal to the `gasLimit` (since 100% of gas is always used).
*  `executionGasUsed` is added as a new field in the transaction receipt.
```

No need to bold every line in a list.

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,236 @@
+* **The `gasUsed` is always equal to the `gasLimit` (since 100% of gas is always used)**
+* Lastly, **we add `executionGasUsed` as a new field in the transaction receipt.
+
+There is a per transaction gas limit of 9 million in L2 and when you compile your contracts for L2, gas usage can increase by more than 5x.
```
```suggestion
There is a per transaction gas limit of 9 million on L2 and when you compile your contracts for L2, gas usage can increase by more than 5x.
```
We should be consistent in this phrasing

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,236 @@
+
+### Nuisance Gas
+
+Nuisance gas is a separate subject from the kind the topics we covered above.
```
```suggestion
Nuisance gas is a separate subject from the topics we covered above.
```

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,236 @@
+
+So, what is _nuisance gas_?
+
+**Nuisance gas exists to **provide an upper bound on the _maximum possible L1 gas_ that a [_fraud proof_](https://research.paradigm.xyz/optimism#fraud-proofs) can cost.**
```
```suggestion
Nuisance gas **provides an upper bound on the _maximum possible L1 gas_ that a [_fraud proof_](https://research.paradigm.xyz/optimism#fraud-proofs) can cost.**
```

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,236 @@
+
+Thus, **we can see that this sender paid roughly `$78.42` for an SNX deposit to L2 OE**, an amount that is equal to what you would expect for an L1 Ethereum gas cost for a deposit.
+
+### Main cost of a deposit
```
This section seems like it's trying to be a "how to read etherscan" tutorial. I assume it's motivated by previous attempts to answer certain questions, but it feels like it goes beyond the scope. I'd try to make it more concise by asking: 
1. What question are we trying to answer?
2. What information do we need to present here? 
3. What knowledge should we take as prerequisite (and provide links for users to gain that knowledge if they don't have it).

maurelian reviewed (changes requested):
I left a lot of little nits here to address. 

More generally, I think the page needs to be a shorter, and more concise. As a reader looking for an answer to a specific question, I'm going to scroll up and down the whole page, and still won't really be sure where to look to get my answer.

I think the way to pare this down is to list out the questions we're trying to answer, and think about what needs to be on this page to answer those questions (and only those questions).

platocrat commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,236 @@
+We'll highlight several important fields from the `Overview` section here that we'll expand on:
+
+* **Interacted With (To):**  Tells us the contract that was interacted with in this transaction.
+* **Transaction Action**: A concise explanation of the overall action that is being made in this transasction. You can use this field as a quick summary of the transaction details, excluding the fee that was paid and timestamps.
```
This is perfect, thanks!

platocrat commented:
> More generally, I think the page needs to be a shorter, and more concise. As a reader looking for an answer to a specific question, I'm going to scroll up and down the whole page, and still won't really be sure where to look to get my answer.
> 

> I think the way to pare this down is to list out the questions we're trying to answer, and think about what needs to be on this page to answer those questions (and only those questions).

I included your changes and cleaned up some stuff.
I also changed the headers into questions and bolded important text.

Both of these changes I think make the page easier to skim.

snario commented:
How is `executionPrice` decided?

platocrat commented:
@snario added a very brief section on this `executionPrice` variable.
Does this new section clarify?

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,196 @@
+
+### How are fees calculated on L2?
+
+Gas fees on L2 are charged in ETH (remember, [L2 ETH is not the same as ETH on L1](http://community.optimism.io/docs/protocol/evm-comparison.html#native-weth)) 
```
This should be https. 
Please use the relative linking format (ie. https://github.com/ethereum-optimism/community-hub/blob/159432ef32b0783422d4961ee9ba56c8df018999/src/docs/protocol/protocol.md#L49). That should also ensure https is used.

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,196 @@
+
+Gas fees on L2 are charged in ETH (remember, [L2 ETH is not the same as ETH on L1](http://community.optimism.io/docs/protocol/evm-comparison.html#native-weth)) 
+
+However, **the gas fee payment itself is equal to to your `gasLimit * gasPrice`**, a calculation that is slightly different from L1 where the fee you are charged is based on `gasUsed`.
```
I know I wanted to be more concise, but if the audience is users here, we should define each of these values.

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
````diff
@@ -0,0 +1,196 @@
+As a developer, it's important to know that `eth_estimateGas` returns:
+
+```
+rollupTxSize * dataPrice + executionPrice * gasUsed
````
Same for these valued: `rollupTxSize`, `dataPrice`, `executionPrice`

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,196 @@
+
+**To minimize gas costs for users, we recommend minimizing your `rollupTxSize`.**
+For example, you can reduce your `uint256` state variables to `uint32` variables wherever possible in your transaction inputs.
+By reducing the size of your state variables, you will likely cut costs to reduce your `rollupTxSize` in exchange for increasing the amount of `gasUsed` for your transaction.
```
State variables or input variables? I'd assume the latter, but I'm not sure.

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,196 @@
+
+Two critical points to remember for this section here are:
+
+1. you should never drop your `eth_gasPrice` or your transaction will simply get rejected on L2, and
```
I think `reduce` is a better term than `drop`. Also 'drop from' what?

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,196 @@
+## 🏦 2. Initiating an L1 Deposit
+
+The most immediate thing you'll likely want to do is deposit ether (ETH) to L2 so that you can interact with contracts and applications while on L2.
+(Note, [ETH on L2 is actually WETH](http://community.optimism.io/docs/protocol/evm-comparison.html#native-weth), but we abstract this for you 🙂.)
```
please use relative link format.

maurelian replied:
```suggestion
(Note, [ETH on L2 is actually WETH](http://community.optimism.io/docs/protocol/evm-comparison.html#native-weth), but we abstract this away for you 🙂.)
```

maurelian commented on src/docs/users/fees-in-optimistic-ethereum.md:
```diff
@@ -0,0 +1,196 @@
+If we look at the `Transaction Action`, we see a deposit was made for roughly `212.35 SNX` to the Synthetix L2 Bridge. Although this field gives us an easy look at the gas cost, _**it's very important to remember that gas costs (i.e. transaction fees) are dependent on the changing gas price and gas used by the transaction!**_
+What this means is that you should really be mindful of the `Gas Price` in Gwei and the `Gas Used by Transaction` for your transactions.
+
+Next, if we look at the `Transaction Fee`, we can see that **this sender paid roughly `$78.42` for an SNX deposit to L2, a gas cost that you would expect for an L1 deposit.**
```
```suggestion
Next, if we look at the `Transaction Fee`, we can see that **this sender paid roughly `$78.42` for an SNX deposit to L2, a gas cost that you would expect for a somewhat complex L1 transaction.**
```

maurelian reviewed (approved):
I made some suggestions, but go ahead and merge when you feel good about it.

platocrat commented on src/docs/protocol/evm-comparison.md:
```diff
@@ -105,6 +103,14 @@ The sequencer assigns its transactions' values for these opcodes based on what i
 
+As a reminder, L1 ETH was created before there were [Ethereum Improvement Proposal (EIP)](https://eips.ethereum.org/) contract standards.
+Inevitably, this has led to some awkward contract interactions on Ethereum, such as having to exchange ETH for [wrapped ether (or WETH)](https://weth.io/) to engage in numerous applications (e.g. this is prominent DeFi).
+While some apps try to obviate the user experience (UX) of having to perform this ETH for WETH exchange/trade (e.g. Rarible does this quite well already), it still results in a suboptimal UX.
```
> Feels awkward calling out a project not everyone will be familiar with.

Should I remove this?

platocrat commented on src/docs/protocol/evm-comparison.md:
```diff
@@ -105,6 +103,14 @@ The sequencer assigns its transactions' values for these opcodes based on what i
+As a reminder, L1 ETH was created before there were [Ethereum Improvement Proposal (EIP)](https://eips.ethereum.org/) contract standards.
+Inevitably, this has led to some awkward contract interactions on Ethereum, such as having to exchange ETH for [wrapped ether (or WETH)](https://weth.io/) to engage in numerous applications (e.g. this is prominent DeFi).
+While some apps try to obviate the user experience (UX) of having to perform this ETH for WETH exchange/trade (e.g. Rarible does this quite well already), it still results in a suboptimal UX.
+Users are still shown that WETH is being used, so the UX is _not completely_ rid of WETH.
```
> Unclear what you mean, where are users shown this?

Should I remove this?
